pkg/controller/registry/grpc/source_test.go

```diff
@@ -61,18 +61,20 @@ func (f *FakeSourceSyncer) sync(state SourceState) {
 		f.History[state.Key] = []connectivity.State{}
 	}
 	f.History[state.Key] = append(f.History[state.Key], state.State)
-	f.expectedEvents -= 1
-	if f.expectedEvents == 0 {
+	if state.State == connectivity.Ready {
+		f.expectedReadies -= 1
+	}
+	if f.expectedReadies == 0 {
```

Comment:
I think I might need more coffee in me this morning but I'm having a hard time groking why this check is needed: it seems like we're transitioning from the test that ensures that we're encountering a specific set of connection state event histories, to a test that ensures we're eventually catching the Ready state and making some inferences about any missing events along the way. Is this the right read, or am I missing something entirely?

Reply:
The test assumed that the sync function is called at every state changes of the grpc connection but it can miss some.  The GRPC connection is started at `conn, err := grpcConnection(address)` in `Add` function and the state change monitoring is started at `if source.Conn.WaitForStateChange(timer, state) {` in `watch` function.  Some state changes can happen between these 2 points so the sync function may not be called for every state changes. 
 The code changes are to make sure that all connection become the ready state and the connection doesn't have any unexpected states in the transition to the ready state.

Reply:
Thanks again for another PR @akihikokuroda!

I am a bit concerned with the proposed changes as there is no guarantee that the catalog is moving through phases as we expect. For example, I may have written the test expecting that the catalog end in the Ready phase but the test could pass if the catalog entered the "ready" phase prior to failing. I see two possible solutions:

1. Update the existing solution to ensure that the expected events occur sequentially and finish in the final expected state.
2. Fix the race condition in which history can be lost (I am unfamiliar with [this code](https://github.com/operator-framework/operator-lifecycle-manager/blob/c6cc1d7f706d355e34b1fc6523693bc13ceb4650/pkg/controller/registry/grpc/source.go#L169-L195) and not sure if this is possible).

Reply:
I would like to point out that I believe that the proposed changes are overall reasonable.

Reply:
After looking at the proposed changes a bit more - its clear that every catalog reaches the "ready" state and that the "captured" history doesn't contain any unexpected states. This seems fine for now, but it'd be great if we could remove the race condition.

Reply:
I am comfortable merging this if we create an issue capturing the race condition.

Reply:
I think my preference is towards fixing the unit test vs. fixing the race condition, but I'd like to track fixing that potential race condition somewhere (e.g. a github issue) so it doesn't get swept under the rug. Any thoughts?

Reply:
OK.  I agree to open an git issue for it.

Reply:
I opened a git issue https://github.com/operator-framework/operator-lifecycle-manager/issues/2537 for the race condition investigation.